File: sources/telegram/h2_pool.py
"""H2 WW/UA chat pool curation. Keep aligned with internal/filter/h2_pool.go."""

from __future__ import annotations

H2_CIS_ARBITRAGE_HANDLES = frozenset(
    {
        "maximaffiliate",
        "zuevaff",
        "zuevchatcpa",
        "traffic_trickster",
        "bakalov_info",
        "seodroplab",
        "po_ushi_v_gambling",
        "frbs_team",
        "g_gate_media",
        "affwriter",
        "m2ensenchannel",
        "cpalenta",
        "cpa_lenta",
    }
)

H2_CIS_ARBITRAGE_SUBSTRINGS = (
    "zuevaff",
    "maximaff",
    "cpalent",
    "cparip",
    "arbitrage_ru",
    "_ru_arb",
    "ru_arbitrage",
)

_WW_BUYER_VOICE = (
    "voluum",
    "redtrack",
    "binom",
    "keitaro",
    "self-hosted",
    "postback",
    "usdt",
    "trc20",
    "tracker",
    "alternative",
    "clickhouse",
)


def _cis_ru_setup_context(lower: str) -> bool:
    return any(
        x in lower
        for x in (
            "россия",
            "russia",
            " ru ",
            " рф ",
            ".ru",
            "+7",
            "сбер",
            "рубл",
        )
    )


def _has_ww_buyer_voice(lower: str) -> bool:
    return any(h in lower for h in _WW_BUYER_VOICE)


def reject_h2_cis_pool(username: str, texts: list[str] | None = None) -> tuple[bool, str]:
    user = username.strip().lstrip("@").lower()
    if user:
        if user in H2_CIS_ARBITRAGE_HANDLES:
            return True, "h2_cis_arbitrage"
        for sub in H2_CIS_ARBITRAGE_SUBSTRINGS:
            if sub in user:
                return True, "h2_cis_username"
    parts = [str(t).strip() for t in (texts or []) if str(t).strip()]
    blob = " ".join(parts).lower()
    if "cpa.rip" in blob and _cis_ru_setup_context(blob) and not _has_ww_buyer_voice(blob):
        return True, "h2_cpa_rip_ru"
    return False, ""

File: internal/filter/h2_pool.go
package filter

import "strings"

// H2 WW/UA pool curation: drop RU/CIS arbitrage channels at discover ingress.
// Keep aligned with sources/telegram/h2_pool.py.

var h2CISArbitrageHandles = map[string]struct{}{
	"maximaffiliate":     {},
	"zuevaff":            {},
	"zuevchatcpa":        {},
	"traffic_trickster":  {},
	"bakalov_info":       {},
	"seodroplab":         {},
	"po_ushi_v_gambling": {},
	"frbs_team":          {},
	"g_gate_media":       {},
	"affwriter":          {},
	"m2ensenchannel":     {},
	"cpalenta":           {},
	"cpa_lenta":          {},
}

var h2CISArbitrageSubstrings = []string{
	"zuevaff",
	"maximaff",
	"cpalent",
	"cparip",
	"arbitrage_ru",
	"_ru_arb",
	"ru_arbitrage",
}

// RejectH2CISPool drops known CIS/RU arbitrage pool noise (SHARDING_SESSION H2).
func RejectH2CISPool(username string, texts ...string) (bool, string) {
	user := strings.ToLower(strings.TrimPrefix(strings.TrimSpace(username), "@"))
	if user != "" {
		if _, ok := h2CISArbitrageHandles[user]; ok {
			return true, "h2_cis_arbitrage"
		}
		for _, sub := range h2CISArbitrageSubstrings {
			if strings.Contains(user, sub) {
				return true, "h2_cis_username"
			}
		}
	}
	var parts []string
	for _, t := range texts {
		t = strings.TrimSpace(t)
		if t != "" {
			parts = append(parts, t)
		}
	}
	blob := strings.ToLower(strings.Join(parts, " "))
	if blob == "" {
		return false, ""
	}
	if strings.Contains(blob, "cpa.rip") && cisRUSetupContext(blob) && !hasWWBuyerVoiceH2(blob) {
		return true, "h2_cpa_rip_ru"
	}
	return false, ""
}

func cisRUSetupContext(lower string) bool {
	return strings.Contains(lower, "россия") ||
		strings.Contains(lower, "russia") ||
		strings.Contains(lower, " ru ") ||
		strings.Contains(lower, " рф ") ||
		strings.Contains(lower, ".ru") ||
		strings.Contains(lower, "+7") ||
		strings.Contains(lower, "сбер") ||
		strings.Contains(lower, "рубл")
}

func hasWWBuyerVoiceH2(lower string) bool {
	hints := []string{
		"voluum", "redtrack", "binom", "keitaro", "self-hosted", "postback",
		"usdt", "trc20", "tracker", "alternative", "clickhouse",
	}
	for _, h := range hints {
		if strings.Contains(lower, h) {
			return true
		}
	}
	return false
}
